### PR TITLE
[#93705416] Create vagrant environment for Jenkins

### PR DESCRIPTION
**Pull request by actionjack**
[Setup a Jenkins Server](https://www.pivotaltracker.com/story/show/93705416)

**What**

This PR is to setup a working [Jenkins](https://jenkins-ci.org/) instance in `vagrant` so we can do local development and testing of the Multi-Cloud PaaS [Continuous Delivery](http://refcardz.dzone.com/refcardz/continuous-delivery-patterns) pipeline.

There are several components to this pull request:
1. https://github.com/alphagov/multicloud-deploy/commit/db9f91bffcac42b9cd088cc2cba729995bf59870 Create a new machine in `vagrant` that will be used as a `Jenkins` master,
2. https://github.com/alphagov/multicloud-deploy/commit/da6e643988be9b5eeb5e936bdbeff22408bb49b7 Use `ansible-galaxy` command to pull down third party ansible playbooks,
3. https://github.com/alphagov/multicloud-deploy/commit/47ff25d2935055e55934b40f66aa26d105819262 Configure `Jenkins` with relevant plugins and a default [DSL seed job](https://wiki.jenkins-ci.org/display/JENKINS/Job+DSL+Plugin),
4. https://github.com/alphagov/multicloud-deploy/commit/74610734fc0e286108325513026bec3cba16c8c6 Create helper [Makefile](http://www.gnu.org/software/make/manual/make.html) to make vagrant environment spin up a simple task.

**How to review this PR**

I've tried to logically split the commits so they are easy to review individually while:
- Keeping a consistent narrative and,
- Not breaking master.

This PR should be reviewed in the following context:

_I would like to:_
- Create a base `vagrant` setup that will be used to to create a `Jenkins` server,
- Get some `ansible playbooks` that will make it easy for me to configure that `Jenkins` server,
- Use `ansible` to configure my Jenkins instance,
- Create a helper script that will make it easy to get the `Jenkins` server up and running.

**Who should review this PR**
- Any one in the core team with the exception of @lucorsel who I paired with.

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    end
+
+    roles.each do |role, hosts|
+      f.write("\n[#{role}]\n")
+      f.write(hosts.join)
+    end
+  end
+end
+
+generate_inventory(HOSTS, INVENTORY_FILE) if %w{up provision}.include?(ARGV[0])
```
Comment (dcarley) [inline Vagrantfile]: I don't think we need to do this if we have a single host and no host-level variables. I think we can use Vagrant's builtin `ansible.groups` instead: http://docs.vagrantup.com/v2/provisioning/ansible.html

File: Vagrantfile [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+      c.vm.provider :vmware_fusion do |v|
+        v.vmx["memsize"] = host[:memory] || MEMORY_DEFAULT
+      end
+    end
+  end
+
+  config.vm.provision :shell, inline: "apt-get purge -qq -y --auto-remove chef puppet"
+
+  config.vm.define HOSTS.last[:name] do |host|
```
Comment (dcarley) [inline Vagrantfile]: We can remove this out block. It was a hack in alphagov/tsuru-ansible@9508525e6d18dce09576fb286c22b4dc313d8cd8 to account for ordering where we had mutlple hosts that depended on one another.

File: Vagrantfile [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    end
+  end
+end
+
+generate_inventory(HOSTS, INVENTORY_FILE) if %w{up provision}.include?(ARGV[0])
+
+Vagrant.configure(2) do |config|
+  config.vm.box = "ubuntu/trusty64"
+  config.ssh.private_key_path = File.expand_path('~/.vagrant.d/insecure_private_key')
+  config.ssh.insert_key = false
```
Comment (dcarley) [inline Vagrantfile]: I think we can omit these two lines and use the defaults. Likewise the `ansible.raw_arguments` below. It was a workaround in alphagov/tsuru-ansible@c2d02f4a1bf5c3a81c0eeee3bc2405ea16e9fc81 for delegation, which we're not using here.

File: Vagrantfile [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+
+  HOSTS.each do |host|
```
Comment (dcarley) [inline Vagrantfile]: We can probably get rid of `HOSTS` too because we only want a single instance.

File: site.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    - name: install unzip
+      apt: name=unzip
+    - name: download terraform
+      get_url: url=https://dl.bintray.com/mitchellh/terraform/terraform_0.4.2_linux_amd64.zip dest=/tmp/terraform.zip mode=0440
+      notify:
+        - unzip terraform.zip to /usr/local/bin
+    - name: copy default dsl seed config.xml to jenkins master
+      copy: src=dsl/config.xml dest=/tmp/config.xml owner=jenkins group=jenkins mode=0644
+    - name: create default dsl seed job
+      shell: "curl -X POST 'http://172.18.10.9.xip.io:8080/createItem?name=default-dsl-job' --data-binary '@config.xml' -H 'Content-Type: text/xml'"
```
Comment (dcarley) [inline site.yml]: Should this IP be `127.0.0.1` so that it works outside of Vagrant?

File: site.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    - geerlingguy.jenkins
+  vars:
+    jenkins_plugins:
+      - git
+      - ssh
+      - job-dsl
+  tasks:
+    - name: install git
+      apt: name=git
+    - name: install maven
```
Comment (dcarley) [inline site.yml]: What is maven used by?

File: site.yml (hunk 5 above)
Comment (actionjack) [inline site.yml]: Good catch missed that!

File: site.yml (hunk 6 above)
Comment (actionjack) [inline site.yml]: You know what I don't even know why I'm installing it, I thought it may have been a dependency for the job DSL but it's optional so I'll take it out.

Comment (dcarley) [inline site.yml]: One less Maven in the world :+1:

**Comment by dcarley** [comment]
When editing `dsl/config.xml` and re-running Ansible it looks like the template job changes but not the Terraform job. Is this expected? Do you have any thoughts about how we'll deploy changes to the job?

**Comment by actionjack** [comment]
@dcarley I've made the changes to the `Vagrantfile` and `site.yml` you suggested.

File: Vagrantfile [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  config.vm.provider :vmware_fusion do |v|
+    v.vmx["memsize"] = MEMORY_DEFAULT
+  end
+
+  config.vm.network "forwarded_port", guest: 8080, host: 8080
+
+  config.vm.provision :shell, inline: "apt-get purge -qq -y --auto-remove chef puppet"
+	config.vm.provision :ansible do |ansible|
+    ansible.groups = { "jenkins-master" => ["default"] }
+ 	  ansible.raw_arguments = "--private-key=~/.vagrant.d/insecure_private_key"
```
Comment (dcarley) [inline Vagrantfile]: The indentation is hard to follow here. Possibly because of some hard-tabs in the original file. Can you reformat as:

``` ruby
  config.vm.provision :shell, inline: "apt-get purge -qq -y --auto-remove chef puppet"
  config.vm.provision :ansible do |ansible|
    ansible.groups = { "jenkins-master" => ["default"] }
    ansible.raw_arguments = "--private-key=~/.vagrant.d/insecure_private_key"
    ansible.verbose = "vv"
    ansible.playbook = "site.yml"
  end
```

Comment (actionjack) [inline Vagrantfile]: Gak! It look fine in vim but less shows it differently

**Comment by actionjack** [comment]
@dcarley :%retab!

**Comment by dcarley** [comment]
LGTM. Two outstanding things are the comment about keys and the bug/feature about updating the job.

**Comment by jimconner** [comment]
The only thing I'd change is from using multiple individual apt steps in the site.yaml to doing it all in one hit...

```
- name: install git
  apt: name=git
- name: install unzip
  apt: name=unzip
```

to:

```
- name: Install required packages
action: apt pkg={{item}} state=installed
with_items:
  - git
  - unzip
```

Scarecely worth it with only two items in the list, but it looks a bit cleaner

**Comment by actionjack** [comment]
@dcarley @jimconner PR updated.

**Comment by jimconner** [comment]
Looks good.

**Comment by actionjack** [comment]
I need to issue a new PR since I can't reopen this one due to a forced push.

This PR has been replaced by https://github.com/alphagov/multicloud-deploy/pull/2

**Comment by annashipman** [comment]
I think someone with admin (i.e. me, Dan or Carl) can reopen a closed pull request but it says the branch has been deleted.
